src/Ledger/Fees.lagda

```diff
@@ -0,0 +1,71 @@
+\begin{code}[hide]
+{-# OPTIONS --safe #-}
+
+open import Ledger.Prelude hiding (_%_; _*_)
+open import Ledger.Abstract
+open import Ledger.Transaction
+
+module Ledger.Fees
+  (txs : _) (open TransactionStructure txs)
+  -- (abs : AbstractFunctions txs) (open AbstractFunctions abs)
+  where
+
+open import Data.Rational using (0ℚ; ℚ; mkℚ+; _*_; floor)
+open import Data.Nat.Coprimality using (1-coprimeTo)
+open import Data.Nat.Induction using (<′-wellFounded)
+open import Data.Nat.Properties using (<⇒<′; ≰⇒>)
+open import Data.Integer using (∣_∣)
+open import Data.Product using (swap)
+open import Induction.WellFounded using (Acc; acc)
+\end{code}
+
+The function \AgdaFunction{scriptsCost} calculates the fee for reference scripts in the transaction
+using a function that is piece-wise linear in the size, where the linear constant multiple grows with each
+\refScriptCostStride bytes.  Thus, the \AgdaFunction{scriptsCost} function depends on the
+\AgdaFunction{scriptsTotalSize} function, which returns an integer that is the total size of
+the reference script in bytes, as well as the following protocol parameters:
+\begin{itemize}
+\item \refScriptCostMultiplier, a rational number, the growth factor or step multiplier that determines how
+much the price per byte increases after each increment;
+\item \refScriptCostStride, an integer, the size in bytes at which the price per byte grows linearly;
+\item \minFeeRefScriptCoinsPerByte, a rational number, the base fee or initial price per byte.
+\end{itemize}
+
+\begin{code}[hide]
+_↑ℚ : ℕ → ℚ
+n ↑ℚ = mkℚ+ n 1 (1-coprimeTo n ∘ swap)
+\end{code}
+\begin{code}
+scriptsCost : (pp : PParams) → ℕ → Coin
+scriptsCost pp scSz with (PParams.refScriptCostStride pp)
+... | 0 = 0  -- This case should never occur; refScriptCostStride should always be > 0.
+... | suc m =
+  scriptsCostAux  0ℚ
+                  minFeeRefScriptCoinsPerByte scSz
+                  (<′-wellFounded scSz)
+\end{code}
+\begin{code}[hide]
+  where
+  open PParams pp
+  multiplier = refScriptCostMultiplier
+  sizeIncrement = suc m
+  sizeIncrementRational = sizeIncrement ↑ℚ
+
+  scriptsCostAux : ℚ → ℚ → (n : ℕ) → Acc _<′_ n → Coin
```

Comment:
Now you've hidden too much, this part needs to be visible. It's a bit annoying that it can't be entirely disentangled from the proof part, but we can't do much about that.

Reply:
Okay, I've added the first couple of lines of `scriptsCostAux`, omitting the details of the proof.  Also, I've put the code inside a labeled figure for consistency, and added a reference to the figure in the paragraph describing the code.